uuid-cluster-dns/README.md

```diff
@@ -0,0 +1,49 @@
+# Using UUIDs for cluster endpoint domains
+
+The current approach for building cluster domains is to include the workload cluster's ID as a subdomain of the management cluster it belongs to, e.g.:
+
+```
+https://api.j9yr2.k8s.gauss.eu-west-1.aws.gigantic.io
+```
+
+This approach has a few drawbacks:
+
+1. The workload cluster is tied to the management cluster, making migration messy
+2. Workload clusters cannot be "promoted" to management clusters as the domain would no longer make sense
+3. Several customers have their own domain set up, leading to inconsistency between MCs
+4. The provider (and region) is baked into the domain
+
+An alternative "flat" DNS structure has been proposed that removes the link between WC and MC. E.g.:
+
+```
+https://api.j9yr2.k8s.[customer name].gigantic.io
+```
+
+While this solves some of the above drawbacks (1, 2, 4) it introduces its own complications:
+
+1. The customer name (or codename) is exposed in the domain
+2. There is no enforced global uniqueness on cluster IDs so if a customer has multiple MCs there is a chance that two workload clusters have the same ID, leading to conflicting domains
+
+## Proposal
+
+The flat DNS structure can be improved to avoid the drawbacks above my leveraging a UUID for the cluster ID rather than a user-providable short string.
+
+E.g.
+
+```
+https://api.4c6aba87-061c-458d-91b6-62ad07979728.k8s.gigantic.io
```

Comment:
What would the impact on automation be? E.g. how could a user easily find this URL in e.g. some automation pipeline?

Reply:
How do they currently find the kubeconfig?

Reply:
But doesn't this have impact beyond the kubeconfig - e.g. for ingress names?

Finding the kubeconfig is usually done through a CLI where it's not a big issue. Maybe I am overthinking it here

Reply:
I think we'd have to place this in the Cluster CR somewhere, e. g. as an annotation. Clients like happa have to expose this to the end user e. g. in order to show how to get access to the Kubernetes endpoint.

Also this is related to a question I had: Should the creator of a cluster be able to specify the DNS name?

Reply:
I'm not sure. What issues specifically are you thinking?

I imagine that if you have a kubeconfig then all other endpoints you need can be queried from the kube API. Not the nicest if doing manually but can be hidden away with some tooling perhaps.

Reply:
Do we have any requirements beyond `opsctl` and `kubectl gs` working?

Reply:
Happa also. But not sure beyond that.

Reply:
@AverageMarcus In happa we explain how to get a workload cluster's ingress base domain. See https://happa.g8s.gollum.westeurope.azure.gigantic.io/organizations/giantswarm/clusters/c68pn/ingress for example:

![image](https://user-images.githubusercontent.com/273727/163997070-d255c505-a480-438d-bbeb-9cf4a0181151.png)

As long as we can derive the base domain from the workload clusters Kubernetes API endpoint, and as long as we can find that one in the Cluster CR, I think we are fine. And, in contrast to what I wrote above regarding an annotation, I'm assuming this will always be found in the cluster CR's `.spec.controlPlaneEndpoint.host` field.

Reply:
Oh that's very nice! :D